### ISSUE TITLE
Provide a method in the event loop to 'process events' or 'do one iteration'

### ISSUE THREAD

**Issue by kawing-chiu** [jonathanslenders/python-prompt-toolkit#475]
Many libs that has an event loop provides a method to ask the event loop to process all pending events and then return, for example, qt's `QCoreApplication::processEvents()`, matplotlib's `canvas.flush_events()`, ipython kernel's `kernel.do_one_iteration()`. Such a method is very convenient when integrating multiple event loops in a big program. Maybe I missed it, I couldn't find something equivalent in prompt-toolkit yet.

**Comment by jonathanslenders** [comment]
You're right. This definitely makes sense.

Right now, the event loop implementation of prompt_toolkit is actually rather minimal. It has just enough functionality to do what prompt_toolkit itself needs. However, we can definitely add more functionality, and something like this makes sense.

We're however not really planning to compete with asyncio or Twisted. So, if ever you need an "advanced" event loop, then it's better to use asyncio, and use the asyncio integration of prompt_toolkit.

Finally, prompt_toolkit 2.0 is coming somewhere during the next few months. This is what I'm spending most of my free time on. This release refactors much of the internal architecture and a few changes to the event loops are included. So, this feature request will have to wait for 2.0.

**Comment by 2sn** [comment]
Is there an example on how to integrate the matplotlib event loop with prompt_toolkit, maybe with asyncio?
All my attempts so far failed, though ipython seems to be an example that it must be possible...

**Comment by jonathanslenders** [comment]
Yes, have a look at this example: https://github.com/jonathanslenders/python-prompt-toolkit/blob/master/examples/prompts/inputhook.py

We support inputhooks, these are functions that are called whenever the prompt_toolkit eventloop is idle. This inputhook is supposed to return whenever ``context.fileno()`` becomes ready.

**Comment by 2sn** [comment]
OK, so you don't know the details of how ipython/matplotlib do these.  I experimented some with their sleep function but that seems to take a lot longer than specified (0.2 s even if 1e-4 s is given) and it does not seem to be cooperating with your use of asyncio.  The stdlib `cmd` package using `readline` seems to work; they must be using some form of threading, though when I search for threads or processes, none show up.  Thanks for your help anyway.  Probably need to ask the matplotlib developers what "magic" they use.
